### PR TITLE
WIP: Nightly builds pyarrow pandas

### PR DESCRIPTION
**Pull request by mganesh1308**

### CODE REVIEW

File: ci/set_numpy_pandas_integration.sh [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+
+set -xeo pipefail
+
+pip-compile ./ci/numpy_pandas_requirements.txt -o ./ci/numpy_pandas_req_pinned.txt
```
Comment (crepererum) [inline ci/set_numpy_pandas_integration.sh]: `ci/numpy_pandas_requirements.txt` isn't that file missing in this PR?

File: ci/set_numpy_pandas_integration.sh [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+echo "Upgrading to  Nightly build of numpy pandas"
+
+PRE_WHEELS="https://7933911d6844c6c53a7d-47bd50c35cd79bd838daf386af554a83.ssl.cf2.rackcdn.com"
+
+pip install --pre --no-deps --upgrade --timeout=180 --no-cache-dir -f $PRE_WHEELS numpy pandas
+
+pip freeze
+
+rm -f ./ci/numpy_pandas_req_pinned.txt
```
Comment (kagharpure) [inline ci/set_numpy_pandas_integration.sh]: I don't think that this line is required.

File: ci/set_pyarrow_integration.sh [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+pip install -r ./ci/pyarrow_req_pinned.txt
+
+echo "Upgrading to Nightly build of pyarrow"
+
+pip install --pre --no-deps --upgrade --timeout=180 --no-cache-dir -f "https://github.com/ursa-labs/crossbow/releases/download/latest" pyarrow
+
+pip freeze
+
+rm -f ./ci/pyarrow_req_pinned.txt
```
Comment (kagharpure) [inline ci/set_pyarrow_integration.sh]: I don't think that this line is required.

File: ci/set_numpy_pandas_integration.sh (hunk 1 above)
Comment (kagharpure) [reply ci/set_numpy_pandas_integration.sh]: It's there.

Comment (mganesh1308) [reply ci/set_numpy_pandas_integration.sh]: ci/numpy_pandas_requirements.txt  is availble in this PR and ./ci/numpy_pandas_req_pinned.txt is temp file.

File: ci/set_numpy_pandas_integration.sh (hunk 2 above)
Comment (mganesh1308) [reply ci/set_numpy_pandas_integration.sh]: sure, will remove if not required..

**Comment by fjetter** [comment]
Ah, this is another PR. @mganesh1308 why create a second PR?

**Comment by mganesh1308** [comment]
> Ah, this is another PR. @mganesh1308 why create a second PR?
my bad since, some issues happened in old PR builds, so created a new one..I will maintain this as final PR and I will update here for further changes!

**Comment by mganesh1308** [comment]
> Didn't I just post a comment with a proposed solution? Anyhow, here it comes again:
> 
> ### ci/compile_requirements.sh
> ```shell
> #!/bin/bash
> set -xeo pipefail
> 
> PIP_COMPILE_OPTIONS=(
>     "--pre"
>     "-f https://7933911d6844c6c53a7d-47bd50c35cd79bd838daf386af554a83.ssl.cf2.rackcdn.com"
> )
> 
> PIP_COMPILE_ARGS=""
> for opt in "${PIP_COMPILE_OPTIONS[@]}";
> do
>     PIP_COMPILE_ARGS=$PIP_COMPILE_ARGS" "$opt
> done
> 
> if [ ! -z ${KARTOTHEK_ARROW_VERSION} ];
> then
>     echo pyarrow==$KARTOTHEK_ARROW_VERSION > kartothek_env_reqs.txt
>     trap 'rm -f kartothek_env_reqs.txt' EXIT
>     pip-compile \
>         $PIP_COMPILE_ARGS \
>         -o requirements-pinned.txt \
>         kartothek_env_reqs.txt \
>         requirements.txt
> else
>     pip-compile \
>         $PIP_COMPILE_ARGS \
>         -o requirements-pinned.txt \
>         requirements.txt
> fi
> 
> pip-compile \
>     $PIP_COMPILE_ARGS \
>     -o test-requirements-pinned.txt \
>     requirements-pinned.txt \
>     test-requirements.txt
> ```
> 
> You'd still need to inject the Options variable in the build matrix but this would allow you to inject the pre-wheels without altering anything majorly

fine @fjetter,
  will check on this.

**Comment by mganesh1308** [comment]
> > Didn't I just post a comment with a proposed solution? Anyhow, here it comes again:
> > ### ci/compile_requirements.sh
> > ```shell
> > #!/bin/bash
> > set -xeo pipefail
> > 
> > PIP_COMPILE_OPTIONS=(
> >     "--pre"
> >     "-f https://7933911d6844c6c53a7d-47bd50c35cd79bd838daf386af554a83.ssl.cf2.rackcdn.com"
> > )
> > 
> > PIP_COMPILE_ARGS=""
> > for opt in "${PIP_COMPILE_OPTIONS[@]}";
> > do
> >     PIP_COMPILE_ARGS=$PIP_COMPILE_ARGS" "$opt
> > done
> > 
> > if [ ! -z ${KARTOTHEK_ARROW_VERSION} ];
> > then
> >     echo pyarrow==$KARTOTHEK_ARROW_VERSION > kartothek_env_reqs.txt
> >     trap 'rm -f kartothek_env_reqs.txt' EXIT
> >     pip-compile \
> >         $PIP_COMPILE_ARGS \
> >         -o requirements-pinned.txt \
> >         kartothek_env_reqs.txt \
> >         requirements.txt
> > else
> >     pip-compile \
> >         $PIP_COMPILE_ARGS \
> >         -o requirements-pinned.txt \
> >         requirements.txt
> > fi
> > 
> > pip-compile \
> >     $PIP_COMPILE_ARGS \
> >     -o test-requirements-pinned.txt \
> >     requirements-pinned.txt \
> >     test-requirements.txt
> > ```
> > 
> > 
> > You'd still need to inject the Options variable in the build matrix but this would allow you to inject the pre-wheels without altering anything majorly
> 
> fine @fjetter,
> will check on this.

@fjetter , this need to update the current compile_requirements.sh(used for default builds) or we need to modify this only for nightly builds(numpy,pandas/pyarrow) which should be in ci/compile_requirements.sh! ? since we are passing $PIP_COMPILE_ARGS in all places and this will install dev versions pandas and numpy in our default builds which we don't want to do right. please clarify.

**Comment by fjetter** [comment]
> I will maintain this as final PR and I will update here for further changes!

I don't really see a reason why you needed to create a new one in the first place but in any case, please close the other one if you want to stick to this one

**Comment by fjetter** [comment]
> this need to update the current compile_requirements.sh(used for default builds) 

Yes, please update the original script to serve all build branches

File: .idea/.gitignore [hunk 4]
```diff
@@ -0,0 +1,2 @@
+# Default ignored files
```
Comment (fjetter) [inline .idea/.gitignore]: Please do not check in your IDE byproducts. See the section [here](https://help.github.com/en/articles/ignoring-files) about global gitignore files for that topic

File: requirements.txt [hunk 5]
```diff
@@ -4,8 +4,7 @@ msgpack>=0.5.2
 # Currently dask and numpy==1.16.0 clash
 numpy!=1.15.0,!=1.16.0
 pandas>=0.23.0
-# pyarrow==0.14.0 breaks kartothek
-pyarrow>=0.11.1, !=0.14.0, <0.14.1 # Keep upper bound pinned until we see non-breaking releases in the future
+pyarrow
```
Comment (fjetter) [inline requirements.txt]: The pinnings are required to be there. The packaged wheels use this information to communicate the compatibility to the users

File: compile_requirements.sh [hunk 6]
```diff
@@ -1,20 +1,59 @@
 #!/bin/bash
 set -xeo pipefail
 
-productionIndex=platform
-developmentIndex=platform_dev
+if [ -z ${KARTOTHEK_NP_PD_NIGHTLY} ]; then
+  KARTOTHEK_NP_PD_NIGHTLY=0
```
Comment (fjetter) [inline compile_requirements.sh]: Please keep indentation at four spaces

File: compile_requirements.sh [hunk 7]
```diff
@@ -1,20 +1,59 @@
...
-    trap 'rm -f kartothek_env_reqs.txt' EXIT
+    echo " KARTOTHEK_ARROW_VERSION Value--->  $KARTOTHEK_ARROW_VERSION"
+    pyarrow_url=$(python ci/get_pyarrow_nightly.py)
+    echo "pyarrow URl --> $pyarrow_url"
+    KARTOTHEK_ARROW_VERSION=$(python -c "print('$pyarrow_url'.split('/')[-1].split('-')[1])")
+    echo pyarrow=="$KARTOTHEK_ARROW_VERSION" > kartothek_env_reqs.txt
+    trap 'rm -f kartothek_env_reqs.txt requirements-pinned.txt' EXIT
+    pip-compile\
+        --pre \
+        -f "$pyarrow_url" \
```
Comment (fjetter) [inline compile_requirements.sh]: Why not use the same mechanism of extending the `PIP_COMPILE_OPTIONS` here?

File: compile_requirements.sh [hunk 8]
```diff
@@ -1,20 +1,59 @@
...
+        -o requirements-pinned.txt \
+        kartothek_env_reqs.txt \
+        requirements.txt\
+
+elif [ ! "${KARTOTHEK_ARROW_VERSION}" = "NIGHTLY" ];
+then
+    echo " KARTOTHEK_ARROW_VERSION Value--->  $KARTOTHEK_ARROW_VERSION"
+    echo pyarrow=="$KARTOTHEK_ARROW_VERSION" > kartothek_env_reqs.txt
+    trap 'rm -f kartothek_env_reqs.txt requirements-pinned.txt' EXIT
```
Comment (fjetter) [inline compile_requirements.sh]: Feels like we have a lot of duplication here. Can you look into rewriting the if/elif statements to use less duplication?

File: .travis.yml [hunk 9]
```diff
@@ -10,23 +10,23 @@ python:
...
   - KARTOTHEK_ARROW_VERSION=0.13.0
+  - KARTOTHEK_ARROW_VERSION=NIGHTLY
+  - KARTOTHEK_NP_PD_NIGHTLY=1
 
 before_install:
-  # https://github.com/JDASoftwareGroup/kartothek/issues/94
   - pip install --upgrade pip==19.1.*
   - pip install pip-tools
-  - ./compile_requirements.sh
+  - travis_wait 8 ./compile_requirements.sh
```
Comment (fjetter) [inline .travis.yml]: The default timeout for travis commands is at 10min, i.e. 10min w/out any output will kill the job but I don't see what would take 10min in the compile requirements step. What's takes so much time here?

File: ci/get_pyarrow_nightly.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import json
+import urllib.request
```
Comment (fjetter) [inline ci/get_pyarrow_nightly.py]: I'd prefer the usage of https://pypi.org/project/requests/2.7.0/

File: ci/get_pyarrow_nightly.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+dict_nightly_builds = {}
+
+with urllib.request.urlopen(
+    "https://api.github.com/repos/ursa-labs/crossbow/releases"
+) as url:
+    data = json.loads(url.read().decode())
```
Comment (fjetter) [inline ci/get_pyarrow_nightly.py]: Is this data stored anywhere? Looks like this script is merely printing stuff

File: compile_requirements.sh [hunk 12]
```diff
@@ -1,20 +1,59 @@
...
+    echo "pyarrow URl --> $pyarrow_url"
+    KARTOTHEK_ARROW_VERSION=$(python -c "print('$pyarrow_url'.split('/')[-1].split('-')[1])")
+    echo pyarrow=="$KARTOTHEK_ARROW_VERSION" > kartothek_env_reqs.txt
+    trap 'rm -f kartothek_env_reqs.txt requirements-pinned.txt' EXIT
+    pip-compile\
+        --pre \
+        -f "$pyarrow_url" \
+        -o requirements-pinned.txt \
+        kartothek_env_reqs.txt \
+        requirements.txt\
```
Comment (fjetter) [inline compile_requirements.sh]: This is probably not valid syntax

**Comment by fjetter** [comment]
For an easier approach on how to pass specific versions to pip-compile, see https://github.com/JDASoftwareGroup/kartothek/pull/112

**Pull request closed without merging**